### PR TITLE
feat(deployments): K8s Deployment+Service for Always workloads (AIRCORE-757 phase 4)

### PR DESCRIPTION
**Pull request by tylersbray**
## Summary

- Implements AIRCORE-757 **phase 4**: `restart_policy: Always` workloads are created as `apps/v1.Deployment` + `v1.Service` (ClusterIP) with in-cluster endpoints (`http://{name}.{namespace}.svc.cluster.local:{port}`).
- Extends K8s status mapping for Deployments: `LOST` when missing, `ImagePullBackOff` → `STARTING`, `CrashLoopBackOff` (≥3 restarts) → `FAILED`.
- Wires `K8sDeploymentBackend` to route Always → `deployments.py`, merge Job + Deployment names in `list_managed_deployment_names`, and attempt both deletes during orphan cleanup.
- Hardens create/delete: verify Deployment ownership before Service creation, roll back Deployment on Service failure, surface config validation errors in `read_status`, and propagate deployment delete failures during orphan cleanup.

**Deferred to later phases:** full `K8sDeploymentConfig` / PodSpec compilation (init containers, multi-container, ConfigMaps — phase 5); custom-namespace listing for managed deployments; RBAC/chart (phase 6); integration tests (phase 7).

## Test plan

- [x] `uv run pytest plugins/nemo-deployments/tests/unit -q` (233 passed)
- [x] Pre-commit hooks on both commits
- [x] Pre-MR review + Bugbot pass on branch changes

## Summary by CodeRabbit

* **New Features**
  * Added Kubernetes “Deployment-backed” support for restart policy **Always**, including service exposure, log retrieval, and full lifecycle handling (create/read/delete).
  * Improved resource discovery by listing managed deployments alongside jobs and consolidating managed name reporting.
* **Bug Fixes**
  * More resilient deletion with best-effort cleanup when Kubernetes resources are missing or partially present.
  * Enhanced status reporting for missing, terminating, and crash-loop scenarios with clearer status messages.
* **Tests**
  * Expanded unit tests for deployment lifecycle, status mapping, and edge-case behaviors (including log and endpoint handling).

### CODE REVIEW

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/status.py [hunk 1]
```diff
@@ -82,3 +85,116 @@ def status_from_job(
         )
 
     return BackendStatusUpdate(status="STARTING", status_message=f"Job {job_name} is pending")
+
+
+def _deployment_is_deleting(deployment: Any) -> bool:
```
Comment (mckornfield) [inline plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/status.py]: super nit
```suggestion
def _is_deployment_deleting(deployment: Any) -> bool:
```

File: plugins/nemo-deployments/tests/unit/backends/k8s/k8s_helpers.py [hunk 2]
```diff
@@ -44,6 +45,90 @@ def sample_deployment(*, config_name: str = "cfg1") -> Deployment:
...
+
+
+def deployment_list_item(*, workspace: str, name: str) -> MagicMock:
+    item = MagicMock()
+    item.metadata.labels = {
+        MANAGED_BY_KEY: MANAGED_BY_LABEL,
+        DEPLOYMENT_WORKSPACE_LABEL: workspace,
+        DEPLOYMENT_NAME_LABEL: name,
+        RESTART_POLICY_LABEL: "Always",
+        CONFIG_NAME_LABEL: "cfg1",
```
Comment (mckornfield) [inline plugins/nemo-deployments/tests/unit/backends/k8s/k8s_helpers.py]: this is dumb, but can these be `config1` or `config-1`, Idk why but I keep seeing cfg as an acronym or something lol

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+
+def app_selector_labels(resource_name: str) -> dict[str, str]:
+    return {APP_LABEL: resource_name}
+
+
+def validate_config_for_deployment(config: DeploymentConfig) -> Container:
+    """Return the sole container spec for a Deployment-backed deployment."""
+    if config.init_containers:
+        raise DeploymentConfigError("init_containers are not supported by the k8s Deployment backend in this phase")
+    if len(config.containers) != 1:
```
Comment (mckornfield) [inline plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py:63]: will things like istio mess with this or no, since they mutate after?

or is this just intermittent until a later state anyhow

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+    resource_name: str,
+    labels: dict[str, str],
+    config: DeploymentConfig,
+    container: Container,
+    workspace: str,
+) -> Any:
+    """Build an ``apps/v1.Deployment`` for create."""
+    k8s = k8s_client_module()
+    selector_labels = app_selector_labels(resource_name)
+    pod_labels = {**selector_labels, **labels}
```
Comment (mckornfield) [inline plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py]: does this have any advantages over ` selector_labels | labels`? maybe 3.9 support?

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+                if deployment_created:
+                    try:
+                        apps_v1.delete_namespaced_deployment(
+                            name=resource_name,
+                            namespace=namespace,
+                            propagation_policy="Background",
+                            _request_timeout=timeout,
+                        )
+                    except ApiException:
+                        pass
```
Comment (mckornfield) [inline plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py]: should these be debug logs instead? I guess this assumes they were already deleted?

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+                return ""
+            return core_v1.read_namespaced_pod_log(
+                name=pod.metadata.name,
+                namespace=namespace,
+                tail_lines=tail,
+                timestamps=True,
+                _request_timeout=timeout,
+            )
+
+        text = await asyncio.to_thread(_logs)
```
Comment (mckornfield) [inline plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py:584]: curious, does the kube client not have asyncio support?

**Review by mckornfield — APPROVED** [review]

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/status.py (hunk 1 above)
Comment (tylersbray) [reply plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/status.py]: Done in 3e25b5a — renamed to `_is_deployment_deleting`.

File: plugins/nemo-deployments/tests/unit/backends/k8s/k8s_helpers.py (hunk 2 above)
Comment (tylersbray) [reply plugins/nemo-deployments/tests/unit/backends/k8s/k8s_helpers.py]: Done in 3e25b5a — k8s test helpers now use `config1` instead of `cfg1`.

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py (hunk 3 above)
Comment (tylersbray) [reply plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py:63]: Mesh sidecars (Istio/Linkerd) inject at admission time and never appear in `DeploymentConfig.init_containers`, so this check does not block them. Status polling reads live pod state, so injection shows up as STARTING until the workload is ready. User-specified init containers are deferred to phase 5; added a comment in `validate_config_for_deployment`.

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py (hunk 4 above)
Comment (tylersbray) [reply plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py]: No 3.9 constraint — repo is `>=3.11`. Switched to `selector_labels | labels` in 3e25b5a.

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py (hunk 5 above)
Comment (tylersbray) [reply plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py]: Good catch — those rollback `except ApiException: pass` blocks now log at DEBUG via `_log_cleanup_ignored` (best-effort cleanup; resource may already be gone). 3e25b5a.

File: plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py (hunk 6 above)
Comment (tylersbray) [reply plugins/nemo-deployments/src/nemo_deployments_plugin/backends/k8s/deployments.py:584]: The official `kubernetes` Python client is sync-only. We wrap calls in `asyncio.to_thread`, matching the phase 3 Job backend (`jobs.py`). Documented in the module docstring; native async client is out of scope for this phase.
